### ISSUE TITLE
Deadlock caused by AbstractAddImportCodeFixProvider attempting to get NuGet package sources

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#39889]
**Version Used**: Latest 16.5 Preview2 dogfood build (29518.163.master)

**Repro steps (pretty consistent repro)**

1. Create a new C# project
2. Add code so you have missing usings. For example, just type out a `Console.Write(...);`, but no `using System;` so you get compiler errors. Place the cursor on the squiggle and click save all.
3. Close VS
4. Restart VS, open the same project
5. Hit Ctrl + Dot - if you do this before the red squiggle shows up, I invariably see this deadlock. 

I think there might be other ways to hit this, but I found this was a pretty consistent repro for me.

Dump shared at: `\\<%internal_share%>\public\mavasani\Deadlock_Ctrl_Dot\devenv.dmp`

**Main thread:**
```
>	Microsoft.CodeAnalysis.Workspaces.dll!Roslyn.Utilities.TaskExtensions.WaitAndGetResult_CanCallOnBackground<System.Collections.Immutable.ImmutableArray<System.__Canon>>(System.Threading.Tasks.Task<System.Collections.Immutable.ImmutableArray<System.__Canon>> task, System.Threading.CancellationToken cancellationToken)	Unknown
 	Microsoft.CodeAnalysis.EditorFeatures.Wpf.dll!Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedActionsSourceProvider.SuggestedActionsSource.GetCodeFixes(Microsoft.CodeAnalysis.Editor.Shared.ITextBufferSupportsFeatureService supportsFeatureService, Microsoft.VisualStudio.Language.Intellisense.ISuggestedActionCategorySet requestedActionCategories, Microsoft.CodeAnalysis.Workspace workspace, Microsoft.CodeAnalysis.Document document, Microsoft.VisualStudio.Text.SnapshotSpan range, System.Threading.CancellationToken cancellationToken)	Unknown
 	Microsoft.CodeAnalysis.EditorFeatures.Wpf.dll!Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedActionsSourceProvider.SuggestedActionsSource.GetSuggestedActions(Microsoft.VisualStudio.Language.Intellisense.ISuggestedActionCategorySet requestedActionCategories, Microsoft.VisualStudio.Text.SnapshotSpan range, System.Threading.CancellationToken cancellationToken)	Unknown
 	Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Language.Intellisense.Implementation.LightBulbSession.TryGetSuggestedActionSets.AnonymousMethod__0(Microsoft.VisualStudio.Utilities.IUIThreadOperationContext c)	Unknown
 	Microsoft.VisualStudio.Editor.Implementation.dll!Microsoft.VisualStudio.Editor.Implementation.VSUIThreadOperationExecutor.Execute(Microsoft.VisualStudio.Utilities.UIThreadOperationExecutionOptions executionOptions, System.Action<Microsoft.VisualStudio.Utilities.IUIThreadOperationContext> action)	Unknown
 	Microsoft.VisualStudio.Editor.Implementation.dll!Microsoft.VisualStudio.Editor.Implementation.VSUIThreadOperationExecutor.Execute(string title, string defaultDescription, bool allowCancellation, bool showProgress, System.Action<Microsoft.VisualStudio.Utilities.IUIThreadOperationContext> action)	Unknown
 	Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.Utilities.UIThreadOperationExecutor.Execute(string title, string defaultDescription, bool allowCancellation, bool showProgress, System.Action<Microsoft.VisualStudio.Utilities.IUIThreadOperationContext> action)	Unknown
 	Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Language.Intellisense.Implementation.LightBulbSession.TryGetSuggestedActionSets(out System.Collections.Generic.IEnumerable<Microsoft.VisualStudio.Language.Intellisense.SuggestedActionSet> actionSets, Microsoft.VisualStudio.Language.Intellisense.ISuggestedActionCategorySet overrideRequestedActionCategories)	Unknown
 	Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Language.Intellisense.Implementation.LightBulbBroker.CreateSession(Microsoft.VisualStudio.Language.Intellisense.ISuggestedActionCategorySet requestedActionCategories, Microsoft.VisualStudio.Text.Editor.ITextView textView, Microsoft.VisualStudio.Text.ITrackingPoint triggerPoint, Microsoft.VisualStudio.Text.ITrackingSpan triggerSpan, System.Collections.Generic.Dictionary<Microsoft.VisualStudio.Text.SnapshotSpan, System.Collections.Generic.IEnumerable<Microsoft.VisualStudio.Language.Intellisense.ISuggestedActionsSource>> subjectSpansAndSources, Microsoft.VisualStudio.Language.Intellisense.ISuggestedActionCategorySet applicableCategories, bool preQuerySuggestedActions, bool trackMouse)	Unknown
```

**Background thread:**
```
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.WaitSynchronouslyCore(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.WaitSynchronously(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread()	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask<bool>.CompleteOnCurrentThread()	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run<bool>(System.Func<System.Threading.Tasks.Task<bool>> asyncMethod, Microsoft.VisualStudio.Threading.JoinableTaskCreationOptions creationOptions)	Unknown
 	NuGet.PackageManagement.VisualStudio.dll!NuGet.PackageManagement.VisualStudio.VSSolutionManager.IsSolutionOpen.get()	Unknown
 	NuGet.PackageManagement.VisualStudio.dll!NuGet.PackageManagement.VisualStudio.VSSettings.ResetSolutionSettingsIfNeeded()	Unknown
 	NuGet.PackageManagement.VisualStudio.dll!NuGet.PackageManagement.VisualStudio.VSSettings.SolutionSettings.get()	Unknown
 	NuGet.PackageManagement.VisualStudio.dll!NuGet.PackageManagement.VisualStudio.VSSettings.GetSection(string sectionName)	Unknown
 	NuGet.Configuration.dll!NuGet.Configuration.PackageSourceProvider.LoadPackageSourceLookup(bool byName, NuGet.Configuration.ISettings settings)	Unknown
 	NuGet.Configuration.dll!NuGet.Configuration.PackageSourceProvider.LoadPackageSources(NuGet.Configuration.ISettings settings, System.Collections.Generic.IEnumerable<NuGet.Configuration.PackageSource> defaultPackageSources)	Unknown
 	NuGet.Configuration.dll!NuGet.Configuration.PackageSourceProvider.LoadPackageSources()	Unknown
 	NuGet.VisualStudio.Implementation.dll!NuGet.VisualStudio.VsPackageSourceProvider.GetSources(bool includeUnOfficial, bool includeDisabled)	Unknown
>	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Packaging.PackageInstallerService.GetPackageSources()	Unknown
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.AddImport.AbstractAddImportCodeFixProvider.RegisterCodeFixesAsync(Microsoft.CodeAnalysis.CodeFixes.CodeFixContext context)	Unknown
 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.CodeFixes.CodeFixService.GetCodeFixesAsync(Microsoft.CodeAnalysis.Document document, Microsoft.CodeAnalysis.Text.TextSpan span, Microsoft.CodeAnalysis.CodeFixes.CodeFixProvider fixer, bool isBlocking, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostic> diagnostics, System.Threading.CancellationToken cancellationToken)	Unknown
 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.CodeFixes.CodeFixService.AppendFixesAsync.AnonymousMethod__2(System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostic> dxs)	Unknown
 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.CodeFixes.CodeFixService.AppendFixesOrConfigurationsAsync.AnonymousMethod__2()	Unknown
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformFunctionAsync<System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.CodeFixes.CodeFix>>(Microsoft.CodeAnalysis.Extensions.IExtensionManager extensionManager, object extension, System.Func<System.Threading.Tasks.Task<System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.CodeFixes.CodeFix>>> function, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.CodeFixes.CodeFix> defaultValue)	Unknown
 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.CodeFixes.CodeFixService.AppendFixesOrConfigurationsAsync<Microsoft.CodeAnalysis.CodeFixes.CodeFixProvider>(Microsoft.CodeAnalysis.Document document, Microsoft.CodeAnalysis.Text.TextSpan fixesSpan, System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.Diagnostics.DiagnosticData> diagnosticsWithSameSpan, bool fixAllForInSpan, Microsoft.CodeAnalysis.PooledObjects.ArrayBuilder<Microsoft.CodeAnalysis.CodeFixes.CodeFixCollection> result, Microsoft.CodeAnalysis.CodeFixes.CodeFixProvider fixer, System.Func<Microsoft.CodeAnalysis.Diagnostic, bool> hasFix, System.Func<System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostic>, System.Threading.Tasks.Task<System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.CodeFixes.CodeFix>>> getFixes, System.Threading.CancellationToken cancellationToken)	Unknown
 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.CodeFixes.CodeFixService.AppendFixesAsync(Microsoft.CodeAnalysis.Document document, Microsoft.CodeAnalysis.Text.TextSpan span, System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.Diagnostics.DiagnosticData> diagnostics, bool fixAllForInSpan, bool isBlocking, Microsoft.CodeAnalysis.PooledObjects.ArrayBuilder<Microsoft.CodeAnalysis.CodeFixes.CodeFixCollection> result, System.Threading.CancellationToken cancellationToken)	Unknown
```

**Comment by mavasani** [comment]
@CyrusNajmabadi @sharwell Has the implementation of `AbstractAddImportCodeFixProvider` always invoked into NuGet package manager to load package sources in RegisterCodeFixes callback? It seems this can also lead to UI delays, not just the deadlock that I encountered. IMO, it would be better to not do so in RegisterCodeFixes code path, but have something else in the background compute the package sources and the RegisterCodeFixes code path should only use already computed package sources, if any.

**Comment by CyrusNajmabadi** [comment]
> but have something else in the background compute the package sources and the RegisterCodeFixes code path should only use already computed package sources, if any.

That is how it is supposed to be working.  We lazily and incrementally compute this in the foreground within roslyn's PackageInstallerService.  The code action is only supposed to use the data that has been computed and is available to use by the BG.  This was hte entire intended design here since our codeactions cannot use the FG thread while processing.

**Comment by mavasani** [comment]
Thanks @CyrusNajmabadi - seems no longer to be the case. I will send a PR to fix this.

**Comment by CyrusNajmabadi** [comment]
This changed in this commit: https://github.com/dotnet/roslyn/commit/9582c433bcd5b93e34bdc53cf915ac78c6314de6#diff-553c36111a99746a31cb38b778531378

Previously this value was free threaded (that was a major point of this class).  The BG thread could read it at any point, and the UI thread would occasionally update it.  It was just POCO data that didn't all out to anything.  In the referenced commit it changed to be a computed call that might end up calling out to an external service from the thread that called it.  This violated the design here, leading to the deadlock you're seeing.

I'm not sure why this really changed.

My recommendation, which was something that was brough up roughly *4-5 years ago* was that the nuget team should just give us a free-threaded API.  this was something they said they would do, but it doesn't look like that was ever followed up on.  So we're still trying to deal with this by needing to do all the UI contortions.

**Comment by mavasani** [comment]
It seems @sharwell did confirm the no UI thread requirement [here](https://github.com/dotnet/roslyn/pull/35364/files#r279900654), but that seems to have regressed or is no longer valid. IMO, even if it were restored, we should avoid doing this computation on background thread from code fix computation.

**Comment by CyrusNajmabadi** [comment]
Note: can we push again to get the nuget apis to be free threaded?  it's really annoying that a modern service in VS is thread affinitized in this manner.  I get that there's nothing we can do about some ancient services.  But nuget is a core part of the entire .net story, and this is a super pain.  It's esp. annoying as they asked if we could implement this feature for them 5+ years ago, and as part of that said that they would provide this API we needed asap.

This impl was never meant to be the one that has lasted, it was always a stop-gap to show we could build this.  

--

Aside, at the time i was pretty vehement that i didn't want this to ship.  I stated explicitly how prototypes become the default that we are then shackled with for years.  This is yet another example of that sort of thing happening.  I think the roslyn team needs to be far more willing to push back on teams and say that we won't ship something incomplete until the teams that want these features themselves build the necessary team components so we can do it in a proper fashion :-/

**Comment by nkolev92** [comment]
Created a NuGet side issue: https://github.com/NuGet/Home/issues/8842

We are planning on adding an async version of all of our sync APIs so that should help the general issues. Won't do much in this specific scenario I think.

Not sure if this specific API can be free threaded since settings can be tied to the solution/ VS SKU, but overall rework to use more async APIs is due :)

**Comment by CyrusNajmabadi** [comment]
> Not sure if this specific API can be free threaded

@sharwell can you help here wrt JTF info to see if this is possible?

**Comment by kartheekp-ms** [comment]
It looks like UI thread is blocked [here](https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs#L364) causing UI delay when NuGet tries to [switch to main thread](https://github.com/NuGet/NuGet.Client/blob/787114e8a9fdbc3dac02399fee9b931f8acb9cff/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs#L309) to interact with COM API. JoinableTaskFactory.Run may help to avoid this deadlock?

**Comment by CyrusNajmabadi** [comment]
> It looks like UI thread is blocked here causing deadlock when NuGet tries to switch to main thread to interact with COM API. JoinableTaskFactory.Run may help to avoid this deadlock?

Note: this is bad no matter waht.  We simply do not want to access the UI thread in these scenarios.  It means that BG work ends up stealing this critical resources from the user.  

Ideally these are just truly free-threaded with no UI thread dependency.

**Comment by nkolev92** [comment]
> Ideally these are just truly free-threaded with no UI thread dependency.

The reason why that's difficult is because NuGet discovers settings based on the solution root.

To discover that we use the DTE.
So in the absolute worst case scenario we get here: 
https://github.com/NuGet/NuGet.Client/blob/787114e8a9fdbc3dac02399fee9b931f8acb9cff/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSSettings.cs#L62

https://github.com/NuGet/NuGet.Client/blob/787114e8a9fdbc3dac02399fee9b931f8acb9cff/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs#L311 and/or
https://github.com/NuGet/NuGet.Client/blob/787114e8a9fdbc3dac02399fee9b931f8acb9cff/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs#L419

**Comment by CyrusNajmabadi** [comment]
> The reason why that's difficult is because NuGet discovers settings based on the solution root.
> To discover that we use the DTE.

Get we can off of that?

**Comment by nkolev92** [comment]
Are you asking whether we can get rid of the solution path dependency?
Unfortunately we can't get rid of the solution path dependency, Discovering the NuGet.Config next to the loaded sln is a basic NuGet feature :/ 

Maybe we can look into refactoring the settings initialization and preempt the solution check so it doesn't happen when trying to read the settings. I don't know if that's possible of the top of my head though. Would need to dig in a little.

**Comment by tmat** [comment]
@nkolev92 If you need to retrieve information about the solution from VS and VS does not provide free-threaded async API then please file an issue on VS platform to provide such API.

**Comment by CyrusNajmabadi** [comment]
> Are you asking whether we can get rid of the solution path dependency?

Yes.  Note: as per @tmat 's post, this means either:

1. using existing mechanisms that woudl allow you to do this.  Or, if they don't exist:
2. filing request for those mechanisms to be created.

Note that '2' was supposed to have happened back in 2016 when the nuget search feature was added.

I discussed this extensively with the nuget API team at the time, esp. around how hard it was for roslyn to use these APIs since they were UI-affinitized.  The communication at the time was that nuget would make this happen.  Indeed, us shipping teh 'add nuget reference' prototype was contingent on this work happening.

It looks like that got lost in the sands of time (which def is understandable), but really should be brought back as it was something that was committed to in exchange for the roslyn work to happen, and it really is one of those things that hurts the code, the product and users.

Thanks!

**Comment by nkolev92** [comment]
> Yes. Note: as per @tmat 's post, this means either:

> 1. using existing mechanisms that woudl allow you to do this. Or, if they don't exist:
> 2. filing request for those mechanisms to be created.

Yeah 2 is definitely the only option we have here.

> It looks like that got lost in the sands of time

Quite possible.

It certainly doesn't cost us much to ask for a feature from another team :) 
Idk how feasible/likely these asks will be because I don't have enough context on the platform side of things.

Myself & @kartheekp-ms are reworking all of NuGet's sync IVs APIs to async, and we'll do see if certain ones can be made free threaded wherever possible.

**Comment by CyrusNajmabadi** [comment]
> Myself & @kartheekp-ms are reworking all of NuGet's sync IVs APIs to async, and we'll do see if certain ones can be made free threaded wherever possible.

Awesome.  Async is a start.  But free threaded is def what we want.  i.e. consider how this API is being used:

Roslyn is doing background analysis to determine what issue there are and how they may be fixed.  This involves seeing unbound identifiers and checking if they could be fixed with nuget.  So we need to determine if we can use nuget.  If this ends up hittin the UI thread (even asynchronously), it will impact the user with (hopefully only micro-) delays.  

Given how choppy VS can feel, this really isn't good.  UI thread should be something avoided as much as can, and *esp.* during bg ephemeral analysis.  

I will point out that this we really shoudln't have shipped the prototype without this.  It allowed us to ship bad experiences to the customer, and basically took off all pressure from teh relevant teams to address the problem.  Had the feature *not* shipped because it was dependent on this, then the work almost certainly would have happened.

**Comment by CyrusNajmabadi** [comment]
@nkolev92 You mentioned: 

> We are planning on adding an async version of all of our sync APIs so that should help the general issues.

Are those APIs available yet?  Note that it's *really* bad that we have APIs here that have a UI thread dependency.  Roslyn itself has to go through many contortions to be able to both use these APIs and not impede the user.  UI thread dependencies are really bad for overall user experience as they end up blocking the UI and causing all those pauses and whatnot that users hate :-/

**Comment by jinujoseph** [comment]
reported [here ](https://developercommunity.visualstudio.com/content/problem/933130/vs-1650-preview-30-ide-hang-during-quick-action-su.html)

**Comment by nkolev92** [comment]
> Are those APIs available yet? Note that it's really bad that we have APIs here that have a UI thread dependency. Roslyn itself has to go through many contortions to be able to both use these APIs and not impede the user. UI thread dependencies are really bad for overall user experience as they end up blocking the UI and causing all those pauses and whatnot that users hate :-/

The async APIs have been delayed. 
An epic issue tracking all of the NuGet extensibility changes: https://github.com/NuGet/Home/issues/8872

You can track the progress for the specific API in this issue at https://github.com/NuGet/Home/issues/8902. 

We can't really commit to free threadedness, each of the APIs will be analyzed at its own merit when the rework happened.

**Comment by tmat** [comment]
I don't think Roslyn should use NuGet APIs directly. The right direction, I believe, is for the Code Fix provider to update the project (add `ProjectReference` via a Project System API) and then let Project System apply the update and restore any new packages.

**Comment by CyrusNajmabadi** [comment]
> I don't think Roslyn should use NuGet APIs directly. 

So we need to talk to nuget to even know if they're enabled and what the user has as their package-sources.  That's where we're hanging.  It's not in the adding of project references.  

Note: i'd also be fine with someone else picking up the responsibility for dealing with adding nuget references.  It's just not the core problem we're dealing with here :-/

**Comment by tmat** [comment]
> So we need to talk to nuget to even know if they're enabled and what the user has as their package-sources.

Do we actually need to? The service that provides the API to NuGet package mapping should tell us "this is the name of the package and a version range where this API is available".
We should take that and pass it to the Project System to figure out if there is either a version upgrade needed or a new ProjectReference needed. 
I don't see where we need to talk to NuGet APIs.

> My personal take is that we should likely pull the feature until we can build it in a proper fashion.

I'd not be against that.

> If there are stakeholders who still want this, then we can allocate resources to getting these apis done :-/

We need to ask the Project System to provide the API.
